source/utf8/core.h

```diff
@@ -297,6 +298,58 @@ namespace internal
         return utf8::internal::validate_next(it, end, ignored);
     }
 
+    //the iterator_base class
+    template <typename octet_iterator>
+    class iterator_base : public std::iterator <std::input_iterator_tag, uint32_t> {
+    protected:
+	   using function = std::function<uint32_t(octet_iterator&, octet_iterator)>;
```

Comment:
Use a template param for this. The type erasure costs quite some performance!

Reply:
Instead of std::function, I could use the following interface
`template<typename T>
struct Next {
virtual ~Next(){}
virtual uint32_t next(T&, T) const = 0;
};`
Both versions (std::function and struct Next) decode a large istringstream (45 millions octets) at the same speed.

I don't see how to use a template param as you suggest. Could you provide some code ?

Reply:
I have written another version with an additional template parameter
`template<typename octet_iterator, typename Next> class iterator_base`
with a private member `Next *func` which is a pointer to function.

Unfortunately, all versions run at the same speed. So it's a matter of taste, which version shall be used.

Reply:
Well the point was NOT to use a function pointer. If you use an additional member template you can use a class like your above `Next` but without the virtual and 2 different classes for unchecked and checked.

However after the discussion in the other PR I have doubts that caching the current cp is the real solution: https://github.com/nemtrif/utfcpp/pull/49#issuecomment-552147536